src/main/java/hudson/remoting/Engine.java

```diff
@@ -849,4 +831,55 @@ private SSLSocketFactory getSSLSocketFactory()
      * @since 2.4
      */
     static final int SOCKET_TIMEOUT = Integer.getInteger(Engine.class.getName()+".socketTimeout",30*60*1000);
+
+    private class EngineJnlpConnectionStateListener extends JnlpConnectionStateListener {
+
+        private final RSAPublicKey publicKey;
```

Comment:
Leave it to me to worry about cryptography choices years before they're a problem. Does this actually need to use an RSA key, or can it be any `PublicKey` type technically? Even if that's usually RSA.

I'm not actually sure about alternative algorithms available through X509 certificates, though later versions of Java have added more support there (if BouncyCastle doesn't cover us already).

Looking further down, I see you had to cast the `PublicKey` to the appropriate subtype anyways. I don't see much in here that's specific to using RSA (besides the fact that that's the algorithm in play here).

Reply:
It's an interesting question, but not terribly relevant for the current PR. We may want to look into something like that as an enhancement in some later PR.